### PR TITLE
[SOLR-13865] Migrate replica routing code to SolrJ

### PR DESCRIPTION
**Pull request by HoustonPutman**
# Tests

Tests have been migrated from solr core.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I am authorized to contribute this code to the ASF and have removed any code I do not have a license to distribute.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: solr/solrj/src/test/org/apache/solr/client/solrj/routing/NodePreferenceRulesComparatorTest.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import org.apache.solr.SolrTestCaseJ4;
+import org.apache.solr.common.cloud.Replica;
+import org.apache.solr.common.cloud.ZkStateReader;
+import org.apache.solr.common.params.ShardParams;
+import org.junit.Test;
+
+public class NodePreferenceRulesComparatorTest extends SolrTestCaseJ4 {
+
+  @Test
+  public void testNodePreferenceRulesComparator() throws Exception {
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/solrj/src/test/org/apache/solr/client/solrj/routing/NodePreferenceRulesComparatorTest.java]: I agree with both of these testing comments. I'm originally just porting everything from Solr Core, and once I get that working, I'll try expanding the tests and breaking them up into smaller units.

File: solr/solrj/src/java/org/apache/solr/client/solrj/routing/ReplicaListTransformerManager.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import org.apache.solr.common.SolrException.ErrorCode;
+import org.apache.solr.common.cloud.NodesSysPropsCacher;
+import org.apache.solr.common.params.CommonParams;
+import org.apache.solr.common.params.ShardParams;
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.common.util.NamedList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReplicaListTransformerManager {
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/solrj/src/java/org/apache/solr/client/solrj/routing/ReplicaListTransformerManager.java]: I don't either. The name `ReplicaListTransformerFactory` was already taken so I picked something related, but i'm open to any and all ideas.

**Comment by HoustonPutman** [comment]
Since @CaoManhDat  originally committed the `NodesSysPropsCacher` code, maybe he has an idea on how to replicate the same testing logic in SolrJ? I don't really want to migrate `TestInjection` to SolrJ, but at this point it seems like it may be cleanest option seeing as it only has one dependency in solr-core. In the meantime I've added a `CommonTestInjection` class in SolrJ that just does the logic for injecting `additionalProps`, but it would be awesome if we could keep the injection logic in just one class.

And given how big this PR already is, I think it's probably better to keep the Streaming Expressions and SolrJ client integrations as a separate PR. That way they can get the discussion they deserve and not get overshadowed by this migration.

**Comment by HoustonPutman** [comment]
Ok, the extraneous streaming changes have been reverted, so all tests should pass now.

Work to still do:
- Lock down naming of classes, `ReplicaListTransformerManager` in particular (Maybe something like `RequestReplicaListTransformerFactory` or `UnitedReplicaListTransformerFactory`
- Split out existing unit tests and expand on them.
- Determine if there is a better test injection solution than `CommonTestInjection`

File: solr/solrj/src/java/org/apache/solr/common/util/CommonTestInjection.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+package org.apache.solr.common.util;
+
+import java.util.Map;
+
+/**
+ * Allows random faults to be injected in running code during test runs across all solr packages.
+ *
+ * Set static strings to "true" or "false" or "true:60" for true 60% of the time.
+ *
+ * All methods are No-Ops unless <code>LuceneTestCase</code> is loadable via the ClassLoader used
```
Comment (tflobbe) [inline solr/solrj/src/java/org/apache/solr/common/util/CommonTestInjection.java]: This is not true at this time

File: solr/solrj/src/java/org/apache/solr/client/solrj/routing/NodePreferenceRulesComparator.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+   * A distributed request is made to the first live server in the URL list.
+   * This means it is just as likely to choose current host as any of the other hosts.
+   * This function makes sure that the cores are sorted according to the given list of preferences.
+   * E.g. If all nodes prefer local cores then a bad/heavily-loaded node will receive less requests from
+   * healthy nodes. This will help prevent a distributed deadlock or timeouts in all the healthy nodes due
+   * to one bad node.
+   *
+   * Optional final preferenceRule is *not* used for pairwise sorting, but instead defines how "equivalent"
+   * replicas will be ordered (the base ordering). Defaults to "random"; may specify "stable".
+   */
```
Comment (tflobbe) [inline solr/solrj/src/java/org/apache/solr/client/solrj/routing/NodePreferenceRulesComparator.java]: This needs to be updated

**Comment by HoustonPutman** [comment]
Thanks for finding the documentation issues.

I think I'm good here, unless you think something needs to be added. Should be easy to plug it into the SolrJ client and Streaming Expressions.

**Pull request merged**